README.md
# AZDropdownMenu

[![Build Status](https://www.bitrise.io/app/1dcd88cfb9a0f14e.svg?token=-Rs7yi0wHvWCFAZzx977JA)](https://www.bitrise.io/app/1dcd88cfb9a0f14e)
[![Version](https://img.shields.io/cocoapods/v/AZDropdownMenu.svg?style=flat)](http://cocoapods.org/pods/AZDropdownMenu)
[![License](https://img.shields.io/cocoapods/l/AZDropdownMenu.svg?style=flat)](http://cocoapods.org/pods/AZDropdownMenu)
[![Platform](https://img.shields.io/cocoapods/p/AZDropdownMenu.svg?style=flat)](http://cocoapods.org/pods/AZDropdownMenu)
[![Language](https://img.shields.io/badge/swift-4-orange.svg)](http://swift.org)

AZDropdownMenu is a simple dropdown menu component that supports Swift.

## Screenshots
Code used in the screencast are included in the bundled sample project.

![screencast](https://cloud.githubusercontent.com/assets/879197/13528338/455c738a-e250-11e5-8671-312aa58a63e1.gif)

## Demo Project

To run the demo project, clone the repo with `git clone https://github.com/Azuritul/AZDropdownMenu.git`, and run `pod install` from the Example directory first.

## Requirements
- iOS 8 or above
- Xcode 8 or above
- Swift 3

## Installation

AZDropdownMenu is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
use_frameworks!
pod 'AZDropdownMenu'
```

Use version 1.1.2 if you don't need Swift 4 support.

## Usage

Declare an array of texts that are served as the item in the menu.
```swift
let titles = ["Action 1", "Action 2", "Action 3"]
```
Then pass the array to the initializer
```swift
let menu = AZDropdownMenu(titles: titles)
```

Set up so that the `UIBarButtonItem` is asso



ated with a function that shows the menu
```
let button = UIBarButtonItem(image: UIImage(named: "menu_image"), style: .Plain, target: self, action: "showDropdown")
navigationItem.leftBarButtonItem = menuButton
```

Calling `public func showMenuFromView(view:UIView)` can then show the menu.
```swift
func showDropdown() {
    if (self.menu?.isDescendantOfView(self.view) == true) {
        self.menu?.hideMenu()
    } else {
        self.menu?.showMenuFromView(self.view)
    }
}
```

The handler `public var cellTapHandler : ((indexPath:NSIndexPath) -> Void)?` would be called
when menu item is tapped. So place code in here to do whatever you want. For example

```swift
menu.cellTapHandler = { [weak self] (indexPath: NSIndexPath) -> Void in
    self?.navigationController?.pushViewController(controller, animated:true)
}
```

#### Create menu with icons
Pass in a AZDropdownMenuItemData in the initializer: `public init(dataSource:[AZDropdownMenuItemData])` and you are good to go.

> Take a look at the sample project in this repository to get more usage of the library.

### Configurable options
Currently AZDropdownMenu can be customized with the following properties. More will come in the future.

- `itemHeight` Row height of the menu item. Default is 44.
- `itemColor` The background color of the menu item. Default is white.
- `itemFontColor` The text color of the menu item. Default is black.
- `itemFontSize` Font size of the menu item. Default is 14.0
- `itemFont` Font used in the menu. Default is 'Helvetica'
- `itemAlignment` The alignment of the menu item. Default is left-aligned.
- `itemSelectionColor` The background color of the menu item while it is tapped. Default is gray.
- `overlayAlpha` The transparency for the background overlay. Default is 0.5
- `overlayColor` Color for the background overlay. Default is black.
- `menuSeparatorStyle` The separator of the menu. Default is single line.
- `menuSeparatorColor` The color of the separator. Default is light gray.
- `itemImagePosition` The position of image in the menu. Can be positioned before or after the text. Default is `.Prefix`.
- `shouldDismissMenuOnDrag` If set to true, menu would be dismissed on drag. Default value is `false`.

## Version
- 1.1.3
   - Update for Swift 4 syntax
   
- 1.1.2
   - Update for Swift 3 syntax

- 1.1.1
   - Update for Swift 2.2's #selector syntax

- 1.1.0
   - Added `shouldDismissMenuOnDrag` option, menu would be dismissed on drag if this option is enabled
   - Modified the sample project with an example using `UITableView`.
   - Added gesture handler for the menu.
   - Extract config struct to a separate file
   - Update readme.
- 1.0.0
   - Added `itemImagePosition` option, now can configure the icon position.
   - Changed icon sets and its corresponding location in the sample project.
   - Modified sample project's appearance a little bit.
   - Modified sample project's structure.
   - Added CHANGELOG.md
- 0.6.1
   - Update readme
- 0.6.0
   - Now can use icons in the menu.
   - Added configuration options: `menuSeparatorColor`, `itemFont`
- 0.5.3
   - Now can configure item's background color while tapped

## Credit
- Icon used in in the sample project is [Designed by Skydesign - Freepik.com](http://www.freepik.com/free-photos-vectors/people)

## Author
Chris Wu (Azuritul), azuritul@gmail.com

## License
AZDropdownMenu is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
### 1.1.3 Release Notes
   - Update for Swift 4
   
### 1.1.2 Release Notes
   - Update for Swift 3

#### 1.1.1 Release Notes
   - Update for Swift 2.2
   
#### 1.1.0 Release Notes
   - Added `shouldDismissMenuOnDrag` option, menu would be dismissed on drag if this option is enabled
   - Modified the sample project with an example using `UITableView`.
   - Added gesture handler for the menu.
   - Extract config struct to a separate file
   - Update readme.

#### 1.0.0 Release Notes
   - Added `itemImagePosition` option, now can configure the icon position.
   - Changed icon sets and its corresponding location in the sample project.
   - Modified sample project's appearance a little bit.
   - Modified sample project's structure.
   - Added CHANGELOG.md
   - Names used in code's license block are made consistent.

#### 0.6.1 Release Notes
  - Update readme

#### 0.6.0 Release Notes

   - Now can use icons in the menu.
   - Added configuration options: `menuSeparatorColor`, `itemFont`

#### 0.5.3 Release Notes
   - Now can configure item's background color while tapped


Example/Tests/Tests.swift
import UIKit
import XCTest
import AZDropdownMenu

class Tests: XCTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }

    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }

}


Pod/Classes/AZDropdownMenuConfig.swift
//
//  AZDropdownMenuConfig.swift
//  AZDropdownMenu
//
//  Created by Chris Wu on 2016/3/3.
//  Copyright (c) 2016 Chris Wu. All rights reserved.
//

import Foundation


struct AZDropdownMenuConfig {
    
    var itemColor : UIColor = UIColor.white
    var itemSelectionColor : UIColor = UIColor.lightGray
    var itemAlignment : AZDropdownMenuItemAlignment = .left
    var itemFontColor : UIColor = UIColor(red: 58/255, green: 58/255, blue: 58/255, alpha: 1.0)
    var itemFontSize : CGFloat = 14.0
    var itemFont : String = "Helvetica"
    var overlayAlpha : CGFloat = 0.5
    var overlayColor : UIColor = UIColor.black
    var menuSeparatorStyle : AZDropdownMenuSeperatorStyle = .singleline
    var menuSeparatorColor : UIColor = UIColor.lightGray
    var itemImagePosition : AZDropdownMenuItemImagePosition = .prefix
    var shouldDismissMenuOnDrag : Bool = false
}

/**
 The separator style of the menu
 
 - Singleline: A solid single line
 - None:       No Separator
 */
public enum AZDropdownMenuSeperatorStyle {
    case singleline, none
}

/**
 The position of image icon in the menu
 
 - Prefix:  Place icon before item title
 - Postfix: Place icon after item title
 */
public enum AZDropdownMenuItemImagePosition {
    case prefix, postfix
}


Example/AZDropdownMenu/AppDelegate.swift
//
//  AppDelegate.swift
//  AZDropdownMenu
//
//  Created by Chris Wu on 01/05/2016.
//  Copyright (c) 2016 Chris Wu. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        let rootViewController = ViewController()
        let nav = UINavigationController(rootViewController: rootViewController)

        self.window = UIWindow(frame: UIScreen.main.bounds)
        self.window?.tintColor = UIColor.black
        self.window!.rootViewController = nav
        self.window!.makeKeyAndVisible()
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {

    }

    func applicationDidEnterBackground(_ application: UIApplication) {

    }

    func applicationWillEnterForeground(_ application: UIApplication) {

    }

    func applicationDidBecomeActive(_ application: UIApplication) {

    }

    func applicationWillTerminate(_ application: UIApplication) {

    }

}


Pod/Classes/AZDropdownMenuBaseCell.swift
//
//  AZDropdownMenuBaseCell.swift
//  Pods
//
//  Created by Chris Wu on 01/15/2016.
//  Copyright (c) 2016 Chris Wu. All rights reserved.
//

import Foundation

open class AZDropdownMenuBaseCell : UITableViewCell, AZDropdownMenuCellProtocol {
    
    open func configureData(_ data: AZDropdownMenuItemData) {
        self.textLabel?.text = data.title
    }

    func configureStyle(_ config: AZDropdownMenuConfig) {
        self.selectionStyle = .none
        self.backgroundColor = config.itemColor
        self.textLabel?.textColor = config.itemFontColor
        self.textLabel?.font = UIFont(name: config.itemFont, size: config.itemFontSize)

        switch config.itemAlignment {
        case .left:
            self.textLabel?.textAlignment = .left
        case .right:
            self.textLabel?.textAlignment = .right
        case .center:
            self.textLabel?.textAlignment = .center
        }
    }
}

protocol AZDropdownMenuCellProtocol {
    func configureData(_ data: AZDropdownMenuItemData)
    func configureStyle(_ configuration:AZDropdownMenuConfig)
}

public enum AZDropdownMenuItemAlignment {
    case left, right, center
}


Example/AZDropdownMenu/ViewController.swift
//
//  ViewController.swift
//  AZDropdownMenu
//
//  Created by Chris Wu on 01/05/2016.
//  Copyright (c) 2016 Chris Wu. All rights reserved.
//

import UIKit
import AZDropdownMenu

class ViewController: UIViewController {

    var rightMenu: AZDropdownMenu?

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.hidesBackButton = false
        navigationController?.navigationBar.isTranslucent = false
        view.backgroundColor = UIColor(red: 80/255, green: 70/255, blue: 66/255, alpha: 1.0)
        title = "AZDropdownMenu"

        /// Add DemoButton 1
        let demoButton1 = buildButton("Demo 1")
        demoButton1.addTarget(self, action: #selector(ViewController.onDemo1Tapped), for: .touchUpInside)
        view.addSubview(demoButton1)

        view.addConstraint(NSLayoutConstraint(item: demoButton1, attribute: .centerX, relatedBy: .equal, toItem: view, attribute: .centerX, multiplier: 1, constant: 0))
        view.addConstraint(NSLayoutConstraint(item: demoButton1, attribute: .centerY, relatedBy: .equal, toItem: view, attribute: .centerY, multiplier: 1, constant: -180))

        ///  Add DemoButton 2
        let demoButton2 = buildButton("Demo 2")
        demoButton2.addTarget(self, action: #selector(ViewController.onDemo2Tapped), for: .touchUpInside)
        view.addSubview(demoButton2)

        view.addConstraint(NSLayoutConstraint(item: demoButton2, attribute: .centerX, relatedBy: .equal, toItem: view, attribute: .centerX, multiplier: 1, constant: 0))
        view.addConstraint(NSLayoutConstraint(item: demoButton2, attribute: .centerY, relatedBy: .equal, toItem: view, attribute: .centerY, multiplier: 1, constant: -80))
    }

    func buildButton(_ title: String) -> UIButton {
        let button = UIButton(type: .system)
        button.backgroundColor = UIColor(red: 80/255, green: 70/255, blue: 66/255, alpha: 1.0)
        button.setTitle(title, for: UIControlState())
        button.layer.cornerRadius = 4.0
        button.setTitleColor(UIColor(red: 233/255, green: 205/255, blue: 193/255, alpha: 1.0), for: UIControlState())
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    }

    @objc func onDemo1Tapped() {
        let controller = DemoViewController1()
        let nv = UINavigationController(rootViewController: controller)
        nv.navigationBar.isTranslucent = false
        nv.navigationBar.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.black]
        self.present(nv, animated: true, completion: nil)
    }

    @objc func onDemo2Tapped() {
        let controller = DemoViewController2()
        let nv = UINavigationController(rootViewController: controller)
        nv.navigationBar.isTranslucent = false
        nv.navigationBar.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.black]
        self.present(nv, animated: true, completion: nil)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

}

/// Example controller that shows the use of menu with UIViewController
class DemoViewController1: UIViewController {

    var rightMenu: AZDropdownMenu?

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor(red: 80/255, green: 70/255, blue: 66/255, alpha: 1.0)

        let cancelButton = UIBarButtonItem(image: UIImage(named: "cancel"), style: .plain, target: self, action: #selector(UIViewController.dismissFromController))
        let rightButton = UIBarButtonItem(image: UIImage(named: "options"), style: .plain, target: self, action: #selector(DemoViewController1.showRightDropdown))

        navigationItem.leftBarButtonItem = cancelButton
        navigationItem.rightBarButtonItem = rightButton

        title = "Demo 1"
        rightMenu = buildDummyDefaultMenu()
    }

    @objc func showRightDropdown() {
        if self.rightMenu?.isDescendant(of: self.view) == true {
            self.rightMenu?.hideMenu()
        } else {
            self.rightMenu?.showMenuFromView(self.view)
        }
    }
}

/// Example that shows the use of menu with UITableViewController
class DemoViewController2: UIViewController, UITableViewDataSource, UITableViewDelegate {

    var rightMenu: AZDropdownMenu?
    var tableView: UITableView = UITableView()

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Demo 2"
        view.backgroundColor = UIColor.white

        let cancelButton = UIBarButtonItem(image: UIImage(named: "cancel"), style: .plain, target: self, action: #selector(UIViewController.dismissFromController))
        let menuButton = UIBarButtonItem(image: UIImage(named: "options"), style: .plain, target: self, action: #selector(DemoViewController2.showRightDropdown))
        tableView.frame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height)
        navigationItem.leftBarButtonItem = cancelButton
        navigationItem.rightBarButtonItem = menuButton

        rightMenu = buildDummyCustomMenu()
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        self.tableView.dataSource = self
        self.tableView.delegate = self
        self.view.addSubview(self.tableView)
    }
    
    @objc
    func showRightDropdown() {
        if self.rightMenu?.isDescendant(of: self.view) == true {
            self.rightMenu?.hideMenu()
        } else {
            self.rightMenu?.showMenuFromView(self.view)
        }
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 16
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if let cell = tableView.dequeueReusableCell(withIdentifier: "cell") {

            cell.textLabel?.text = "TestCell \(indexPath.row)"
            return cell
        }
        return UITableViewCell()
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.tableView.deselectRow(at: indexPath, animated: true)
        let nextVC = UIViewController()
        nextVC.view.backgroundColor = UIColor.white
        nextVC.title = "VC \(indexPath.row)"
        self.navigationController?.show(nextVC, sender: self)
    }
}

// MARK: - Extension for holding custom methods
extension UIViewController {

    /**
     Create dummy menu with default settings

     - returns: The dummy menu
     */
    fileprivate func buildDummyDefaultMenu() -> AZDropdownMenu {
        let leftTitles = ["Item 1", "Item 2", "Item 3", "Item 4", "Item 5", "Item 6"]
        let menu = AZDropdownMenu(titles: leftTitles)
        menu.itemFontSize = 16.0
        menu.itemFontName = "Helvetica"
        menu.cellTapHandler = { [weak self] (indexPath: IndexPath) -> Void in
            self?.pushNewViewController(leftTitles[indexPath.row])
        }

        return menu
    }

    /**
     Create dummy menu with some custom configuration

     - returns: The dummy menu
     */
    fileprivate func buildDummyCustomMenu() -> AZDropdownMenu {
        let dataSource = createDummyDatasource()
        let menu = AZDropdownMenu(dataSource: dataSource )
        menu.cellTapHandler = { [weak self] (indexPath: IndexPath) -> Void in
            self?.pushNewViewController(dataSource[indexPath.row])
        }
        menu.itemHeight = 70
        menu.itemFontSize = 14.0
        menu.itemFontName = "Menlo-Bold"
        menu.itemColor = UIColor.white
        menu.itemFontColor = UIColor(red: 55/255, green: 11/255, blue: 17/255, alpha: 1.0)
        menu.overlayColor = UIColor.black
        menu.overlayAlpha = 0.3
        menu.itemAlignment = .center
        menu.itemImagePosition = .postfix
        menu.menuSeparatorStyle = .none
        menu.shouldDismissMenuOnDrag = true
        return menu
    }

    fileprivate func pushNewViewController(_ title: String) {
        let newController = UIViewController()
        newController.title = title
        newController.view.backgroundColor = UIColor.white
        DispatchQueue.main.async(execute: {
            self.show(newController, sender: self)
        })
    }

    fileprivate func pushNewViewController(_ item: AZDropdownMenuItemData) {
        self.pushNewViewController(item.title)
    }

    fileprivate func createDummyDatasource() -> [AZDropdownMenuItemData] {
        var dataSource: [AZDropdownMenuItemData] = []
        dataSource.append(AZDropdownMenuItemData(title:"Action With Icon 1", icon:UIImage(imageLiteralResourceName: "1_a")))
        dataSource.append(AZDropdownMenuItemData(title:"Action With Icon 2", icon:UIImage(imageLiteralResourceName: "2_a")))
        dataSource.append(AZDropdownMenuItemData(title:"Action With Icon 3", icon:UIImage(imageLiteralResourceName: "3_a")))
        dataSource.append(AZDropdownMenuItemData(title:"Action With Icon 4", icon:UIImage(imageLiteralResourceName: "4_a")))
        dataSource.append(AZDropdownMenuItemData(title:"Action With Icon 5", icon:UIImage(imageLiteralResourceName: "5_a")))
        return dataSource
    }

    @objc func dismissFromController() {
        self.dismiss(animated: true, completion: nil)
    }
}


Pod/Classes/AZDropdownMenu.swift
//
//  AZDropdownMenu.swift
//  AZDropdownMenu
//
//  Created by Chris Wu on 01/05/2016.
//  Copyright (c) 2016 Chris Wu. All rights reserved.
//

import UIKit

open class AZDropdownMenu: UIView {

    fileprivate let DROPDOWN_MENU_CELL_KEY : String = "MenuItemCell"

    /// The dark overlay behind the menu
    fileprivate let overlay:UIView = UIView()
    fileprivate var menuView: UITableView!

    /// Array of titles for the menu
    fileprivate var titles = [String]()

    /// Property to figure out if initial layout has been configured
    fileprivate var isSetUpFinished : Bool

    /// The handler used when menu item is tapped
    open var cellTapHandler : ((_ indexPath:IndexPath) -> Void)?

    // MARK: - Configuration options

    /// Row height of the menu item
    open var itemHeight : Int = 44 {
        didSet {
            let menuFrame = CGRect(origin: CGPoint(x: 0,y :0), size: CGSize(width: frame.size.width, height: menuHeight))
            self.menuView.frame = menuFrame
        }
    }

    /// The color of the menu item
    open var itemColor : UIColor = UIColor.white {
        didSet {
            self.menuConfig?.itemColor = itemColor
        }
    }

    /// The background color of the menu item while being tapped
    open var itemSelectionColor : UIColor = UIColor.lightGray {
        didSet {
            self.menuConfig?.itemSelectionColor = itemSelectionColor
        }
    }

    /// The font of the item
    open var itemFontName : String = "Helvetica" {
        didSet {
            self.menuConfig?.itemFont = itemFontName
        }
    }

    /// The text color of the menu item
    open var itemFontColor : UIColor = UIColor(red: 140/255, green: 134/255, blue: 125/255, alpha: 1.0) {
        didSet {
            self.menuConfig?.itemFontColor = itemFontColor
        }
    }

    /// Font size of the menu item
    open var itemFontSize : CGFloat = 14.0 {
        didSet {
            self.menuConfig?.itemFontSize = itemFontSize
        }
    }

    /// The alpha for the background overlay
    open var overlayAlpha : CGFloat = 0.5 {
        didSet {
            self.menuConfig?.overlayAlpha = self.overlayAlpha
        }
    }

    /// Color for the background overlay
    open var overlayColor : UIColor = UIColor.black {
        didSet {
            self.overlay.backgroundColor = self.overlayColor
            self.menuConfig?.overlayColor = self.overlayColor
        }
    }

    open var menuSeparatorStyle:AZDropdownMenuSeperatorStyle = .singleline {
        didSet {
            switch menuSeparatorStyle {
                case .none:
                    self.menuView.separatorStyle = .none
                    self.menuConfig?.menuSeparatorStyle = .none
                case .singleline:
                    self.menuView.separatorStyle = .singleLine
                    self.menuConfig?.menuSeparatorStyle = .singleline
            }
        }
    }

    open var menuSeparatorColor:UIColor = UIColor.lightGray {
        didSet {
            self.menuConfig?.menuSeparatorColor = self.menuSeparatorColor
            self.menuView.separatorColor = self.menuSeparatorColor
        }
    }

    /// The text alignment of the menu item
    open var itemAlignment : AZDropdownMenuItemAlignment = .left {
        didSet {
            switch itemAlignment {
                case .right:
                    self.menuConfig?.itemAlignment = .right
                case .left:
                    self.menuConfig?.itemAlignment = .left
                case .center:
                    self.menuConfig?.itemAlignment = .center
            }
        }
    }

    /// The image position, default to .Prefix.  Image will be displayed after item's text if set to .Postfix
    open var itemImagePosition : AZDropdownMenuItemImagePosition = .prefix {
        didSet {
            switch itemImagePosition {
            case .prefix:
                self.menuConfig?.itemImagePosition = .prefix
            case .postfix:
                self.menuConfig?.itemImagePosition = .postfix
            }
        }
    }

    open var shouldDismissMenuOnDrag : Bool = false

    fileprivate var calcMenuHeight : CGFloat {
        get {
            return CGFloat(itemHeight * itemDataSource.count)
        }
    }

    fileprivate var menuHeight : CGFloat {
        get {
            return (calcMenuHeight > frame.size.height) ? frame.size.height : calcMenuHeight
        }
    }

    fileprivate var initialMenuCenter : CGPoint = CGPoint(x: 0, y: 0)
    fileprivate var itemDataSource : [AZDropdownMenuItemData] = []
    fileprivate var reuseId : String?
    fileprivate var menuConfig : AZDropdownMenuConfig?

    // MARK: - Initializer
    public init(titles:[String]) {
        self.isSetUpFinished = false
        self.titles = titles
        for title in titles {
            itemDataSource.append(AZDropdownMenuItemData(title: title))
        }
        self.menuConfig = AZDropdownMenuConfig()
        super.init(frame:UIScreen.main.bounds)
        self.accessibilityIdentifier = "AZDropdownMenu"
        self.backgroundColor = UIColor.clear
        self.alpha = 0.95
        self.translatesAutoresizingMaskIntoConstraints = false
        initOverlay()
        initMenu()
    }

    public init(dataSource:[AZDropdownMenuItemData]) {
        self.isSetUpFinished = false
        self.itemDataSource = dataSource
        self.menuConfig = AZDropdownMenuConfig()
        super.init(frame:UIScreen.main.bounds)
        self.accessibilityIdentifier = "AZDropdownMenu"
        self.backgroundColor = UIColor.clear
        self.alpha = 0.95
        self.translatesAutoresizingMaskIntoConstraints = false
        initOverlay()
        initMenu()
    }

    required public init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - View lifecycle
    override open func layoutSubviews() {
        if isSetUpFinished == false {
            setupInitialLayout()
        }
    }

    fileprivate func initOverlay() {
        let frame = UIScreen.main.bounds
        overlay.frame = CGRect(origin: CGPoint(x: frame.origin.x,y :frame.origin.y), size: CGSize(width: frame.size.width, height: frame.size.height))
        overlay.backgroundColor = self.overlayColor
        overlay.accessibilityIdentifier = "OVERLAY"
        overlay.alpha = 0
        overlay.isUserInteractionEnabled = true
        let touch : UIGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(AZDropdownMenu.overlayTapped))
        overlay.addGestureRecognizer(touch)
        let panGesture  = UIPanGestureRecognizer(target: self, action: #selector(AZDropdownMenu.handlePan(gestureRecognizer:)))
        panGesture.delegate = self
        overlay.addGestureRecognizer(panGesture)
        addSubview(overlay)
    }

    fileprivate func initMenu() {
        let frame = UIScreen.main.bounds
        let menuFrame = CGRect(origin: CGPoint(x: 0,y :0), size: CGSize(width: frame.size.width, height: menuHeight))

        menuView = UITableView(frame: menuFrame, style: .plain)
        menuView.isUserInteractionEnabled = true
        menuView.rowHeight = CGFloat(itemHeight)
        if self.reuseId == nil {
            self.reuseId = DROPDOWN_MENU_CELL_KEY
        }
        menuView.dataSource = self
        menuView.delegate = self
        menuView.isScrollEnabled = false
        menuView.accessibilityIdentifier = "MENU"
        menuView.separatorColor = menuConfig?.menuSeparatorColor
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(AZDropdownMenu.handlePan(gestureRecognizer:)))
        panGesture.delegate = self
        menuView.addGestureRecognizer(panGesture)
        addSubview(menuView)
    }

    fileprivate func setupInitialLayout() {

        let height = NSLayoutConstraint(item: self, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: UIScreen.main.bounds.height)
        let width = NSLayoutConstraint(item: self, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: UIScreen.main.bounds.width)

        addConstraints([height, width])
        isSetUpFinished = true

    }

    fileprivate func animateOvelay(_ alphaValue: CGFloat, interval: Double, completionHandler: (() -> Void)? ) {
        UIView.animate(
            withDuration: interval,
            animations: {
                self.overlay.alpha = alphaValue
            }, completion: { (finished: Bool) -> Void in
                if let completionHandler = completionHandler {
                    completionHandler()
                }
            }
        )
    }

    @objc func overlayTapped() {
        hideMenu()
    }

    //MARK: - Public methods to control the menu

    /**
    Show menu

    - parameter view: The view to be attached by the menu, ex. the controller's view
    */
    open func showMenuFromView(_ view:UIView) {

        view.addSubview(self)

        animateOvelay(overlayAlpha, interval: 0.4, completionHandler: nil)
        menuView.reloadData()
        UIView.animate(
            withDuration: 0.2,
            delay:0,
            usingSpringWithDamping: 0.9,
            initialSpringVelocity: 0.6,
            options:[],
            animations: {
                self.frame.origin.y = view.frame.origin.y
                }, completion: { (finished : Bool) -> Void in
                self.initialMenuCenter = self.menuView.center
            }
        )
    }

    open func showMenuFromRect(_ rect:CGRect) {
        let window = UIApplication.shared.keyWindow!

        let menuFrame = CGRect(origin: CGPoint(x: 0,y :rect.origin.y), size: CGSize(width: frame.size.width, height: menuHeight))

        self.menuView.frame = menuFrame

        window.addSubview(self)

        animateOvelay(overlayAlpha, interval: 0.4, completionHandler: nil)
        menuView.reloadData()
        UIView.animate(
            withDuration: 0.2,
            delay:0,
            usingSpringWithDamping: 0.9,
            initialSpringVelocity: 0.6,
            options:[],
            animations: {
                self.frame.origin.y = rect.origin.y
            }, completion: { (finished : Bool) -> Void in
                self.initialMenuCenter = self.menuView.center
            }
        )
    }

    open func hideMenu() {

        animateOvelay(0.0, interval: 0.1, completionHandler: nil)

        UIView.animate(
            withDuration: 0.3, delay: 0.1,
            options: [],
            animations: {
                self.frame.origin.y = -1200
            },
            completion: { (finished: Bool) -> Void in
                self.menuView.center = self.initialMenuCenter
                self.removeFromSuperview()
            }
        )
    }
}


// MARK: - UITableViewDataSource
extension AZDropdownMenu: UITableViewDataSource {
    @available(iOS 2.0, *)
    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if let cell = getCellByData() {
            let item = itemDataSource[indexPath.row]
            if let config = self.menuConfig {
                cell.configureStyle(config)
            }
            cell.configureData(item)
            cell.layoutIfNeeded()
            return cell
        }
        return UITableViewCell()
    }


    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return itemDataSource.count
    }

    func getCellByData() -> AZDropdownMenuBaseCell? {
        if let _ = itemDataSource.first?.icon {
            return AZDropdownMenuDefaultCell(reuseIdentifier: DROPDOWN_MENU_CELL_KEY, config: self.menuConfig!)
        } else {
            return AZDropdownMenuBaseCell(style: .default, reuseIdentifier: DROPDOWN_MENU_CELL_KEY)
        }
    }
}


// MARK: - UITableViewDelegate
extension AZDropdownMenu: UITableViewDelegate {

    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath as IndexPath, animated:true)
        cellTapHandler?(indexPath as IndexPath)
        if let cell = tableView.cellForRow(at: indexPath as IndexPath) {
            cell.backgroundColor = itemSelectionColor
        }
        
        hideMenu()
    }
    
    public func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        if let cell = tableView.cellForRow(at: indexPath as IndexPath) {
            cell.backgroundColor = itemColor
        }
    }
    
    public func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return CGFloat(itemHeight)
    }

}

// MARK: - UIGestureRecognizerDelegate
extension AZDropdownMenu: UIGestureRecognizerDelegate {

    @objc public func handlePan(gestureRecognizer: UIPanGestureRecognizer) {
        guard self.shouldDismissMenuOnDrag == true else {
            return
        }

        if gestureRecognizer.isKind(of: UIPanGestureRecognizer.self) {
            if let touchedView = gestureRecognizer.view, touchedView == self.menuView {
                let translationView = gestureRecognizer.translation(in: self)
                switch gestureRecognizer.state {
                    case .changed:
                        let center = touchedView.center
                        let targetPoint = center.y + translationView.y
                        let newLocation = targetPoint < initialMenuCenter.y ? targetPoint : initialMenuCenter.y
                        touchedView.center = CGPoint(x: center.x,y :newLocation)
                        gestureRecognizer.setTranslation(CGPoint(x: 0,y :0), in: touchedView)
                    case .ended:
                        if touchedView.center.y < initialMenuCenter.y {
                            hideMenu()
                        }
                    default:break
                }
            }
        }
    }
}


/**
 *  Menu's model object
 */
public struct AZDropdownMenuItemData {

    public let title:String
    public let icon:UIImage?

    public init(title:String) {
        self.title = title
        self.icon = nil
    }

    public init(title:String, icon:UIImage) {
        self.title = title
        self.icon = icon
    }
}
